### PR TITLE
feat: torch cache-able dataset, with sampling support

### PR DESCRIPTION
**Pull request by eddyxu**
Pytorch Dataset

* Work with PyTorch pipelines
* Local filesystem backed cache
* Faster pseudo random samplings. 
* Support project projection and filterings.

### CODE REVIEW

File: python/python/lance/sampler.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    """
+    if isinstance(dataset, (str, Path)):
+        dataset = lance.dataset(dataset)
+
+    if isinstance(columns, str):
+        columns = [columns]
+
+    if n >= len(dataset):
+        # Dont have enough data in the dataset. Just do a full scan
+        dataset.to_batches(columns=columns, batch_size=batch_size)
```
Comment (AyushExel) [inline python/python/lance/sampler.py:124]: This should have a return/yeild here?

Comment (AyushExel) [reply python/python/lance/sampler.py:124]: And maybe also add this condition in test_sampler, where. n>len(ds)

**Review by AyushExel — COMMENTED** [review]
One small question

**Review by wjones127 — COMMENTED** [review]
The cache looks useful. I don't know how useful the sampling functions are given they do nothing to randomize the order in which the dataset is iterated over, and ML users seem to care a lot about randomization.

**Comment by eddyxu** [comment]
>  how useful the sampling functions are given they do nothing to randomize the order in which the dataset is iterated over, and ML users seem to care a lot about randomization.

This can be done with shuffle in torch.dataset.DataLoader or `tf.data.pipeline`.

File: python/python/lance/cache.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            )
+        else:
+            raise ValueError(f"Unsupported cache type: {type(cache)}")
+        self.cache_file = None
+        self.stream = stream
+
+    def __del__(self):
+        if self.cache_dir is not None:
+            self.cache_dir.cleanup()
+            self.cache_dir = None
```
Comment (chebbyChefNEQ) [inline python/python/lance/cache.py:46]: I don't this this is garueeteed to be called since this is only called during GC.

Comment (chebbyChefNEQ) [reply python/python/lance/cache.py:46]: maybe we can use `import atexit` instead?

File: python/python/lance/cache.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    def __iter__(self):
```
Comment (chebbyChefNEQ) [inline python/python/lance/cache.py:48]: nit: do we wannt check and make sure the second stream doesn't iter faster than the first stream?

File: python/python/lance/cache.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        if self.cache_file is None:
+            # First iteration.
+            writer: pa.ipc.RecordBatchFileWriter = None
+            for batch in self.stream:
+                if writer is None:
+                    self.cache_file = Path(self.cache_dir.name) / "cache.arrow"
+                    writer = pa.ipc.new_stream(str(self.cache_file), batch.schema)
+                writer.write(batch)
```
Comment (chebbyChefNEQ) [inline python/python/lance/cache.py:56]: does this append or overwrite?

File: python/python/lance/cache.py (hunk 2 above)
Comment (eddyxu) [reply python/python/lance/cache.py:46]: `atexit` is strictly later than GC?

File: python/python/lance/cache.py (hunk 4 above)
Comment (eddyxu) [reply python/python/lance/cache.py:56]: this is append to a stream

https://arrow.apache.org/docs/python/generated/pyarrow.ipc.RecordBatchStreamWriter.html#pyarrow.ipc.RecordBatchStreamWriter.write

File: python/python/lance/cache.py (hunk 2 above)
Comment (chebbyChefNEQ) [reply python/python/lance/cache.py:46]: GC isn't garueeteed to happen on an object. But atexit is

File: python/python/lance/cache.py (hunk 3 above)
Comment (eddyxu) [reply python/python/lance/cache.py:49]: Ok, added one.

File: python/python/lance/cache.py (hunk 2 above)
Comment (eddyxu) [reply python/python/lance/cache.py]: Ok, add ti both to `atext` and `__exit__`

File: python/python/lance/sampler.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/sampler.py]: Good catch. Wrote too much rust recently

**Review by wjones127 — APPROVED** [review]
